packages/services/src/providers/spot-live/rpc-spot.provider.ts

```diff
@@ -0,0 +1,775 @@
+      if (chainFilters) {
+        chainFilters.push(filter);
+      }
+    }
+
+    if (protocolFilters.length > 0) {
+      const chainsWithFilters = Array.from(
+        this.protocolFiltersByChain.keys(),
+      ).join(", ");
+      this.logger.info(
+        `[RpcSpotProvider] Protocol filtering enabled for chains: ${chainsWithFilters}`,
+      );
+    } else {
+      this.logger.info(
+        `[RpcSpotProvider] No protocol filtering - all DEXs allowed`,
+      );
+    }
+  }
+
+  /**
+   * Get provider name for logging and debugging
+   * @returns Provider name combining spot provider type with underlying RPC provider
+   */
+  getName(): string {
+    return `RPC Direct (${this.rpcProvider.getName()})`;
+  }
+
+  /**
+   * Check if provider is healthy and responsive
+   * @returns True if provider is operational
+   */
+  async isHealthy(): Promise<boolean> {
+    return await this.rpcProvider.isHealthy();
+  }
+
+  /**
+   * Get all trades for a wallet since a given time
+   * Detects swaps from transfer patterns and applies protocol filtering if configured
+   * @param walletAddress Wallet address to monitor
+   * @param since Start time or block number for scanning
+   * @param chains Array of chains to scan
+   * @returns Array of detected on-chain trades
+   */
+  async getTradesSince(
+    walletAddress: string,
+    since: Date | number,
+    chains: string[],
+  ): Promise<OnChainTrade[]> {
+    const startTime = Date.now();
+    const maskedAddress = this.maskAddress(walletAddress);
+
+    // Validate inputs
+    if (chains.length === 0) {
+      this.logger.warn(
+        `[RpcSpotProvider] No chains provided for trade scanning`,
+      );
+      return [];
+    }
+
+    const firstChain = chains[0];
+    if (!firstChain) {
+      this.logger.warn(`[RpcSpotProvider] Empty chains array`);
+      return [];
+    }
+
+    const sinceBlock =
+      typeof since === "number"
+        ? since
+        : await this.dateToBlockNumber(since, firstChain);
+
+    // Add Sentry breadcrumb for debugging
+    Sentry.addBreadcrumb({
+      category: "spot.rpc",
+      message: `Trades request`,
+      level: "info",
+      data: {
+        walletAddress: maskedAddress,
+        chains,
+        sinceBlock,
+      },
+    });
+
+    this.logger.debug(
+      {
+        wallet: maskedAddress,
+        chains,
+        sinceBlock,
+      },
+      `[RpcSpotProvider] Fetching trades`,
+    );
+
+    const allTrades: OnChainTrade[] = [];
+
+    for (const chain of chains) {
+      try {
+        // Fetch all transfers with pagination support
+        // Alchemy limits responses to 1000 results per request
+        const allTransfers: AssetTransfersWithMetadataResult[] = [];
+        let pageKey: string | undefined;
+
+        do {
+          const transfersResponse = await this.rpcProvider.getAssetTransfers(
+            walletAddress,
+            chain as SpecificChain,
+            sinceBlock,
+            pageKey || "latest",
```

Comment:
I believe this is the signature of getAssetTransfers from rpc.ts:
```
 getAssetTransfers(
    walletAddress: string,
    chain: SpecificChain,
    fromBlock: number | string,
    toBlock: number | string,
  )
```

Not sure ig pagekey = toblock

Reply:
Ended up fixing this in a downstream PR: https://github.com/recallnet/js-recall/pull/1596